### PR TITLE
Implement "Safe" slice for all log event parsing

### PR DESCRIPTION
**Pull request by samsondav**

### CODE REVIEW

**Comment by kaleofduty** [comment]
Is there a way we could introduce a dedicated type "UntrustedBytes"?

File: core/utils/utils.go [hunk 1]
```diff
@@ -515,3 +515,16 @@ func Uint256ToHex(n *big.Int) (string, error) {
 	}
 	return common.BigToHash(n).Hex(), nil
 }
+
+// SafeByteSlice returns an error on out of bounds access to a byte array, where a
+// normal slice would panic instead
+func SafeByteSlice(ary []byte, start int, end int) (slice []byte, err error) {
+	defer func() {
```
Comment (kaleofduty) [inline core/utils/utils.go]: I'd prefer a different solution where we don't use recover, but instead do a bounds check.

Comment (j16r) [reply core/utils/utils.go]: Strong agree on this one.

Comment (samsondav) [reply core/utils/utils.go]: Out of interest, why?

Comment (samsondav) [reply core/utils/utils.go]: I originally had a bounds check version but there were a lot of different options (e.g. start greater than end, negative indexes) and it felt safer just to recover from panic.

Comment (j16r) [reply core/utils/utils.go]: A defer will rescue any kind of panic, some of which may not be related to your failure, the code should be simpler `if x > len(arr) { return err }` right? Also panics cause a bit of a performance impact, due to stack unwinding and adding handlers to the stack for every invocation of he defer.

Comment (samsondav) [reply core/utils/utils.go]: Not that simple.

What if start is negative? What if start > end? There are a bunch of different things that can go wrong here. It feels safer to simply recover.

File: core/services/vrf/vrf_coordinator_solidity_crosscheck_test.go [hunk 2]
```diff
@@ -189,7 +204,7 @@ func TestRandomnessRequestLog(t *testing.T) {
 	assert.Equal(t, jobID, log.JobID, "VRFCoordinator logged different JobID from randomness request!")
 	assert.Equal(t, coordinator.consumerContractAddress, log.Sender, "VRFCoordinator logged different requester address from randomness request!")
 	assert.True(t, equal(fee, (*big.Int)(log.Fee)), "VRFCoordinator logged different fee from randomness request!")
-	parsedLog, err := ParseRandomnessRequestLog(chainlink_eth.Log(log.Raw.Raw))
+	parsedLog, err := ParseRandomnessRequestLog(toCLEthLog(log.Raw.Raw))
```
Comment (samsondav) [inline core/services/vrf/vrf_coordinator_solidity_crosscheck_test.go]: Note that with this custom type, implicit type casts no longer work.

**Comment by j16r** [comment]
`@samsondav WIP: Add type 'UntrustedBytes' `

**Comment by samsondav** [comment]
@j16r good catch, not it's not a WIP but you may like to review it separately. Have changed the commit title.

**Comment by samsondav** [comment]
Thanks @j16r

**Pull request merged**